c++/src/kj/refcount.h

```diff
@@ -102,6 +105,86 @@ Own<T> Refcounted::addRefInternal(T* object) {
   return Own<T>(object, *refcounted);
 }
 
+// =======================================================================================
+// Atomic (thread-safe) refcounting
+//
+// Warning: Atomic ops are SLOW.
+
+class AtomicRefcounted: private kj::Disposer {
+public:
+  virtual ~AtomicRefcounted() noexcept(false);
+
+  inline bool isShared() const { return __atomic_load_n(&refcount, __ATOMIC_ACQUIRE) > 1; }
+
+private:
+  mutable uint refcount = 0;
+
+  bool addRefWeakInternal() const;
+
+  void disposeImpl(void* pointer) const override;
+  template <typename T>
+  static kj::Own<T> addRefInternal(T* object);
+  template <typename T>
+  static kj::Own<const T> addRefInternal(const T* object);
+
+  template <typename T>
+  friend kj::Own<T> atomicAddRef(T& object);
+  template <typename T>
+  friend kj::Own<const T> atomicAddRef(const T& object);
+  template <typename T>
+  friend kj::Maybe<kj::Own<const T>> atomicAddRefWeak(const T& object);
+  template <typename T, typename... Params>
+  friend kj::Own<T> atomicRefcounted(Params&&... params);
+};
+
+template <typename T, typename... Params>
+inline kj::Own<T> atomicRefcounted(Params&&... params) {
+  return AtomicRefcounted::addRefInternal(new T(kj::fwd<Params>(params)...));
+}
+
+template <typename T>
+kj::Own<T> atomicAddRef(T& object) {
+  KJ_IREQUIRE(object.AtomicRefcounted::refcount > 0, "Object not allocated with kj::refcounted().");
+  return AtomicRefcounted::addRefInternal(&object);
+}
+
+template <typename T>
+kj::Own<const T> atomicAddRef(const T& object) {
+  KJ_IREQUIRE(object.AtomicRefcounted::refcount > 0, "Object not allocated with kj::refcounted().");
+  return AtomicRefcounted::addRefInternal(&object);
+}
+
+template <typename T>
+kj::Maybe<kj::Own<const T>> atomicAddRefWeak(const T& object) {
+  // Try to addref an object whose refcount could have already reached zero in another thread, and
+  // whose destructor could therefore already have started executing. The destructor must contain
+  // some synchronization that guarantees that said destructor has not yet completed when
```

Comment:
Does this refer to the atomic operations in `~AtomicRefcounted()`, or does the derived class need to implement some sort of synchronization as well? By its usage in later commits, I presume the answer is the former -- nevertheless I had a moment of confusion here.

Reply:
`atomicAddRefWeak()` actually isn't used in this PR. It was used in Workers at one point. The requirement is actually with regards to the derived class.

The idea is: You certainly cannot call `atomicAddRefWeak()` if the object's destructor has *finished*, since then you're passing in an invalid reference. But the whole point of `atomicAddRefWeak()` is to handle the case where the object's destructor may have *started*. So how could you possibly be in a situation where you know the destructor hasn't finished, but you don't know whether it has started? Well, this must be because you are currently holding a lock, and the object's destructor also tries to take that lock. So if the destructor has started, then that thread is probably blocked waiting for the lock that you currently hold. Probably once the destructor takes that lock, it's going to go null-out whatever pointer it is that led you to the object in the first place.

Weak pointers are hairy, but basically this is half of a weak pointer implementation...

Reply:
I get it now! Hairy indeed -- thanks for the explanation.